File: ws/nacato/u2/ws14ProductPrice/model/product_model.py
import json
import os

class ProductModel:
    def __init__(self, json_file="products.json"):
        self.json_file = json_file
        self.vat_percentage = 0.15
        self.verify_or_create_catalog()

    def verify_or_create_catalog(self):
        """Creates the JSON file with 5 default items if it does not exist."""
        if not os.path.exists(self.json_file):
            default_items = [
                {"id": 1, "product": "Laptop", "price": 850.00},
                {"id": 2, "product": "Desktop PC", "price": 1200.00},
                {"id": 3, "product": "Optical Mouse", "price": 25.50},
                {"id": 4, "product": "16GB RAM Stick", "price": 65.00},
                {"id": 5, "product": "24-inch Monitor", "price": 180.00}
            ]
            self.save_catalog(default_items)

    def save_catalog(self, product_list):
        with open(self.json_file, 'w', encoding='utf-8') as f:
            json.dump(product_list, f, indent=4, ensure_ascii=False)

    def read_catalog(self):
        """Returns the available products from the JSON file."""
        try:
            with open(self.json_file, 'r', encoding='utf-8') as f:
                return json.load(f)
        except (json.JSONDecodeError, FileNotFoundError):
            return []

    def calculate_item_values(self, base_price):
        """Calculates 15% VAT and the total for a single product."""
        vat = base_price * self.vat_percentage
        total = base_price + vat
        return round(vat, 2), round(total, 2)

    def calculate_list_totals(self, chosen_products):
        """Calculates subtotal, VAT, and grand total for the cart."""
        subtotal = sum(p['price'] for p in chosen_products)
        vat = sum(p['vat'] for p in chosen_products)
        total = sum(p['total'] for p in chosen_products)
        return round(subtotal, 2), round(vat, 2), round(total, 2)

File: ws/nacato/u2/ws15ProductPriceJava/src/main/java/ec/edu/espe/ProductPriceJava/model/ProductModel.java
package ec.edu.espe.ProductPriceJava.model;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author Angie Nacato, Error 404, @ESPE
 */
public class ProductModel {
    private final String jsonFile;
    private final double vatPercentage;
    private final Gson gson;

    public ProductModel() {
        this.jsonFile = "products.json";
        this.vatPercentage = 0.15;
        this.gson = new GsonBuilder().setPrettyPrinting().create();
        this.verifyOrCreateCatalog();
    }

    public final void verifyOrCreateCatalog() {
        File file = new File(jsonFile);
        if (!file.exists()) {
            List<Product> defaultItems = new ArrayList<>();
            defaultItems.add(new Product(1, "Laptop", 850.00));
            defaultItems.add(new Product(2, "Desktop PC", 1200.00));
            defaultItems.add(new Product(3, "Optical Mouse", 25.50));
            defaultItems.add(new Product(4, "16GB RAM Stick", 65.00));
            defaultItems.add(new Product(5, "24-inch Monitor", 180.00));
            saveCatalog(defaultItems);
        }
    }

    public void saveCatalog(List<Product> productList) {
        try (FileWriter writer = new FileWriter(jsonFile)) {
            gson.toJson(productList, writer);
        } catch (IOException e) {
            System.err.println("Error saving catalog: " + e.getMessage());
        }
    }

    public List<Product> readCatalog() {
        File file = new File(jsonFile);
        if (!file.exists()) return new ArrayList<>();
        
        try (FileReader reader = new FileReader(file)) {
            Type listType = new TypeToken<ArrayList<Product>>(){}.getType();
            List<Product> catalog = gson.fromJson(reader, listType);
            return catalog != null ? catalog : new ArrayList<>();
        } catch (IOException e) {
            return new ArrayList<>();
        }
    }

    public double[] calculateItemValues(double basePrice) {
        double vat = basePrice * this.vatPercentage;
        double total = basePrice + vat;
        return new double[]{
            Math.round(vat * 100.0) / 100.0, 
            Math.round(total * 100.0) / 100.0
        };
    }

    public double[] calculateListTotals(List<Product> chosenProducts) {
        double subtotalAcumulado = 0.0;
        double vatAcumulado = 0.0;
        double totalAcumulado = 0.0;

        for (Product p : chosenProducts) {
            subtotalAcumulado += p.getPrice();
            vatAcumulado += p.getVat();
            totalAcumulado += p.getTotal();
        }

        double finalSubtotal = Math.round(subtotalAcumulado * 100.0) / 100.0;
        double finalVat = Math.round(vatAcumulado * 100.0) / 100.0;
        double finalTotal = Math.round(totalAcumulado * 100.0) / 100.0;

        return new double[]{finalSubtotal, finalVat, finalTotal};
    }
}
